### PR TITLE
Feature/single wire chip interface

### PR DESCRIPTION
**Pull request by PeterHeckel-TEK**
To support the ecc608 swi variant, the following changes have been made to allow the option to specify the communication protocol at build time using features. 

The functions that interface with the i2c / uart driver have been moved to a new transport.rs file which varies based on feature definition.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,8 @@ repository = "https://github.com/helium/ecc608-linux-rs"
 readme = "README.md"
 
 [dependencies]
-i2c-linux = "0"
+i2c-linux = {version = "0" }
```
Comment (madninja) [inline Cargo.toml]:
```suggestion
i2c-linux = "0"
```

File: src/command.rs [hunk 2]
```diff
@@ -182,7 +182,7 @@ impl EccCommand {
     }
 
     pub fn bytes_into(&self, bytes: &mut BytesMut) {
-        bytes.put_slice(&[0x03, 0x00]);
+        bytes.put_slice(&[0x00, 0x00]);
```
Comment (madninja) [inline src/command.rs]: why this change?

File: src/command.rs [hunk 3]
```diff
@@ -239,18 +239,18 @@ impl EccCommand {
         bytes.put_u16_le(crc(&bytes[1..]))
     }
 
-    pub fn duration(&self) -> Duration {
+    pub fn duration(&self, swi_interface: bool ) -> Duration {
```
Comment (madninja) [inline src/command.rs]: why not dispatch to the abstraction layer instead of using a bool here?

File: src/ecc.rs [hunk 4]
```diff
@@ -174,78 +163,45 @@ impl Ecc {
...
-            buf.clear();
+            buf.clear();         
             command.bytes_into(&mut buf);
+            
+            self.transport.send_wake()?;
 
-            self.send_wake();
-            thread::sleep(WAKE_DELAY);
+            match self.transport.protocol {
+                TransportProtocol::I2c => {buf[0] = ATCA_I2C_COMMAND_FLAG}
```
Comment (madninja) [inline src/ecc.rs]: this can be done much more concisely with a `buf.put_u8(self.transport.protocol.into())`. And yes that means you have to implement a `From<TransportProtocol> for u8`

File: src/ecc.rs [hunk 5]
```diff
@@ -174,78 +163,45 @@ impl Ecc {
...
+            if let Err(_err) = self.transport.send_recv_buf(delay, &mut buf) {
                 if retry == retries {
                     break;
                 } else {
                     continue;
                 }
             }
 
-            let response = EccResponse::from_bytes(&buf[..])?;
+            let response = EccResponse::from_bytes(&buf[..]);
```
Comment (madninja) [inline src/ecc.rs]: where did the error response go? use `inspect` to do the send_sleep bit and keep the `?` at the end o f this?

File: src/ecc.rs [hunk 6]
```diff
@@ -174,78 +163,45 @@ impl Ecc {
...
 
-            self.send_wake();
-            thread::sleep(WAKE_DELAY);
+            match self.transport.protocol {
+                TransportProtocol::I2c => {buf[0] = ATCA_I2C_COMMAND_FLAG}
+                TransportProtocol::Swi => {buf[0] = ATCA_SWI_COMMAND_FLAG}
+            }
 
-            if let Err(_err) = self.send_recv_buf(command.duration(), &mut buf) {
+            let delay = match self.transport.protocol {
```
Comment (madninja) [inline src/ecc.rs]: use something more like `let delay = self.transport.duration(command)`

File: src/transport.rs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        
+                let _ = self.swi_port.as_mut().unwrap().write(&sleep_encoded);
+                thread::sleep( SWI_BIT_SEND_DELAY * 8);
+            },
+        }
+    }
+
+    pub fn send_recv_buf(&mut self, delay: Duration, buf: &mut BytesMut) -> Result {
+        match self.protocol {
+            TransportProtocol::I2c => {
```
Comment (madninja) [inline src/transport.rs]: Implement these on a `TransportProtocol` instead of matching on protocol all over the place?

**Review by madninja — COMMENTED** [review]
Initial set of comments . You'll want to run `cargo fmt`  to get the code laid out right and run a clippy pass `cargo +nightly clippy -- -Dclippy::all` and resolve all the warnings/errors.

File: src/ecc.rs (hunk 5 above)
Comment (PeterHeckel-TEK) [reply src/ecc.rs]: Reason I tossed the `?` error response out is due to our earlier conversation of considering CRC errors recoverable. With this implementation the `match response` handles the error the same but will retry the command if there was a CRC error and there are still retries permitted

File: src/command.rs (hunk 2 above)
Comment (PeterHeckel-TEK) [reply src/command.rs]: While this change isn't directly necessary I guess, it was done due to the i2c chips having a different leading command byte compared to the swi chips (0x03 vs 0x77). While we could just keep the i2c byte as a default here, I thought it would be better to keep it as 0's until the operation on line 172 of ecc.rs where the respective command byte is placed at the start of the send buffer. Ultimately it doesn't matter, so this can stay or go depending on your preference.

File: src/transport.rs (hunk 7 above)
Comment (PeterHeckel-TEK) [reply src/transport.rs]: Could you elaborate on how this would be done? I'm unsure of how implementing on the enum would get away from matching? Wouldn't you still have to match within the `TransportProtocol` implementation?

**Review by PeterHeckel-TEK — COMMENTED** [review]
@madninja Thank you for your suggestions, all the comments I haven't responded to I will implement changes in a coming commit. Others have justifications / further questions.

File: src/ecc.rs (hunk 5 above)
Comment (PeterHeckel-TEK) [reply src/ecc.rs]: Thinking further about this... This could be resolved by making a `CRCError` field in the `EccError` enum and return `Ok(Self::Error(EccError::CRCError))` similarly to how the other errors are handled in `from_bytes()` in command.rs

File: src/command.rs (hunk 2 above)
Comment (madninja) [reply src/command.rs]: Would it not work better to actually ask the transport layer for the preamble byte(s) right here instead of overwriting it somewhere later on?

File: src/command.rs [hunk 8]
```diff
@@ -123,6 +122,8 @@ pub enum EccError {
     /// watchdog timer will expire. The system must reset the watchdog timer by
     /// entering the idle or sleep modes.
     WatchDogError,
+    /// Crc in the message doesn't match the calculated Crc
```
Comment (madninja) [inline src/command.rs]:
```suggestion
    /// Crc in the message does not match the calculated Crc
```

File: src/ecc.rs [hunk 9]
```diff
@@ -176,11 +165,14 @@ impl Ecc {
         for retry in 0..retries {
             buf.clear();
             command.bytes_into(&mut buf);
+            
+            self.transport.send_wake()?;
 
-            self.send_wake();
-            thread::sleep(WAKE_DELAY);
+            buf[0] = self.transport.protocol.into();
```
Comment (madninja) [inline src/ecc.rs]: I recommend using the actual BufMut trait here since `buf` is a `BytesMut`.. i.e. `buf.put_u8(..)`

File: src/transport.rs (hunk 7 above)
Comment (madninja) [reply src/transport.rs]: instead of having a `TransportProtocol` _and_ an `EccTransport`, I would suggest having an enum type for the latter:

```rust
pub(crate) enum EccTransport {
    I2c(I2cTransport),
   Swi(SwiTransport),
}
```

And then have structs for `I2cTransport` and `SwiTransport` that implement the same functions. The top level enum then implements the same api surface but dispatches to the right inner struct. Yes you still have to match on the enum, but it would _only_ happen in the top level EccTransport, leaving each individual nested type to keep just the state it needs and the specific code for that transport without needing to special case code or mix state

File: src/transport.rs [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+        //Because Tx line is linked with Rx line, all sent msgs are returned on the Rx line and must be cleared from the buffer
+        let mut clear_rx_line = BytesMut::new();
+        clear_rx_line.resize(send_size, 0);
+        let _ = self.swi_port.as_mut().unwrap().read_exact( &mut clear_rx_line );
+
+        Ok(())
+    }
+
+    fn recv_swi_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(2) };
```
Comment (madninja) [inline src/transport.rs]: unsafe should really not be needed. use one fo the other BytesMut functions like resize or reserve?

File: src/transport.rs [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+        buf[1] = 0xFF;
```
Comment (madninja) [inline src/transport.rs:219]: what is `buf[0]`  assumed to be here? You may not want to assume anything here?

File: src/transport.rs [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+            return Err(Error::timeout());
+        }
+        unsafe { buf.set_len(count) };
+        if let Err(_err) = self.decode_swi_to_uart(&mut buf[1..count]) {
+            return Err(Error::timeout());
+        }
+        Ok(())
+    }
+
+    fn encode_uart_to_swi(&mut self, uart_msg: &BytesMut ) -> BytesMut {
```
Comment (madninja) [inline src/transport.rs]: You shouldn't really use a BytesMut just fo manipulate a byte, just take a u8 and return one?

File: src/transport.rs [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        let _ = self.swi_port.as_mut().unwrap().read_exact( &mut clear_rx_line );
+
+        Ok(())
+    }
+
+    fn recv_swi_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(2) };
+        buf[1] = 0xFF;
+
+        let mut transmit_flag = BytesMut::new();
```
Comment (madninja) [inline src/transport.rs]: Recommend not using BytesMut for simple known sized arrays or single bytes as I think is the case here.

File: src/transport.rs (hunk 12 above)
Comment (PeterHeckel-TEK) [reply src/transport.rs]: This isn't single byte manipulation, this is to encode an array of bytes. You pass in the entire message you want to send and the function returns a new array of bytes with the SWI encoding.

File: src/transport.rs (hunk 11 above)
Comment (PeterHeckel-TEK) [reply src/transport.rs:219]: `buf[0]` isn't assumed to be anything here as we don't care about it's value. When we send the transmit flag, it will be the first thing in the uart rx buffer (as the tx/rx lines are shorted), so when we try to read the subsequent incoming message, `buf[0]` will always have the transmit flag in it. Hence why we discard it later on line 211

File: src/command.rs (hunk 2 above)
Comment (PeterHeckel-TEK) [reply src/command.rs]: Yeah I agree but this object has no access to the transport layer data, ie what protocol is being used. But I will alter this to just put `[0x00]` and place the transport flag in the buf before this function is called.

**Review by PeterHeckel-TEK — COMMENTED** [review]
Thank you @madninja for your comments, see explanations / justifications below

File: src/transport.rs [hunk 14]
```diff
@@ -165,41 +166,83 @@ impl EccTransport {
...
             flags: ReadFlags::NO_START,
         };
-        self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [read_msg])?;
+        self.port.i2c_transfer(&mut [read_msg])?;
+        Ok(())
+    }
+}
+
+impl SwiTransport {
+    fn init( path: &str ) -> Result<Self> {
```
Comment (madninja) [inline src/transport.rs]: These functions are usually called `new`

File: src/transport.rs [hunk 15]
```diff
@@ -165,41 +166,83 @@ impl EccTransport {
...
+        let port = serialport::new(path, SWI_DEFAULT_BAUDRATE)
+        .data_bits(serialport::DataBits::Seven)
+        .parity(serialport::Parity::None)
+        .stop_bits(serialport::StopBits::One)
+        .timeout(Duration::from_millis(50))
+        .open().unwrap_or_else(|e| {
+            eprintln!("Failed to open serial port. Error: {}", e);
```
Comment (madninja) [inline src/transport.rs]: Do not exit the process, return an error and let the caller deal with it

File: src/transport.rs [hunk 16]
```diff
@@ -78,82 +44,117 @@ impl EccTransport {
...
+    pub fn put_command_flag( &self ) -> u8 {
+        match self {
+            Self::I2c(_) => ATCA_I2C_COMMAND_FLAG,
+            Self::Swi(_) => ATCA_SWI_COMMAND_FLAG,
+        }
+    }
+}
+
+impl I2cTransport {
+    fn init( path: &str, address: u16 ) -> Result<Self> {
```
Comment (madninja) [inline src/transport.rs]: rename to `new`

File: src/transport.rs (hunk 11 above)
Comment (madninja) [reply src/transport.rs:219]: in which case why not do a put_slice or two put_u8's here? instead of resizing?

File: src/transport.rs [hunk 17]
```diff
@@ -221,15 +264,15 @@ impl EccTransport {
         Ok(())
     }
 
-    fn encode_uart_to_swi(&mut self, uart_msg: &BytesMut ) -> BytesMut {
-        
+    fn encode_uart_to_swi(&mut self, uart_msg: &[u8] ) -> BytesMut {
+
         let mut bit_field = BytesMut::new();
```
Comment (madninja) [inline src/transport.rs]: use BytesMut::with_capacity instead of new followed by reserve?

**Review by madninja — COMMENTED** [review]
Almost there!

You'll want to run cargo fmt to get the code laid out right and run a clippy pass cargo +nightly clippy -- -Dclippy::all and resolve all the warnings/errors.

**Comment by madninja** [comment]
I'm working to update gateway-mfr-rs to use a gateway-rs like URI approach to indicate which transport/driver to use avoid having to hack up special cases

File: src/transport.rs (hunk 11 above)
Comment (PeterHeckel-TEK) [reply src/transport.rs:219]: It needs to be sized to be 2 bytes because when we call `self.decode_swi_to_uart` , it will only read the number of bytes equal to the size of the buffer. Having it be only a 2 byte buffer means that we can read the first transfer flag as well as the size of the message. Once we know the size of the message, the buffer is resized again to match the message size and the rest of the message is read. If we use put_slice or put_u8 we are just appending to the end of the buffer and the buffer size is unknown and we may read more of the message than we intend to

**Review by PeterHeckel-TEK — COMMENTED** [review]
Just the one note @madninja

File: src/error.rs [hunk 18]
```diff
@@ -8,7 +8,8 @@ pub enum Error {
     Timeout,
     #[error("ecc error {:?}", .0)]
     Ecc(crate::command::EccError),
-
+    #[error("serial port error")]
+    SerialPort(#[from] serialport::Error),
```
Comment (madninja) [inline src/error.rs:12]: 1. Does this not need a cfg conditional, since that crate is only used when SWI is active?
2. To avoid needing the cfg condition you could translate the error into an IoError?

File: src/transport.rs (hunk 11 above)
Comment (madninja) [reply src/transport.rs:219]: Gotcha, ok

**Review by madninja — COMMENTED** [review]
Just one question around the SerialPort error

File: src/error.rs (hunk 18 above)
Comment (madninja) [reply src/error.rs:12]: Oh SWI Is no optional anymore, I see
